### PR TITLE
IntersectionObserverPolyfill wrapper for amp-a4a refresh.

### PR DESCRIPTION
**Pull request by glevitzky**
These changes add a wrapper class to enclose `IntersectionObserverPolyfill`. The main purpose of the wrapper is to force the observer to update its internal state periodically (which the polyfill does not do on its own), and thus provide consistent updates on the elements it monitors. This is accomplished by effectively overriding the `viewportCallback` function of the element to invoke `tick` on the observer.

**Note**: test-refresh-manager.js was renamed to test-refresh.js. All new tests are under `describe('RefreshIntersectionObserverWrapper', () => { ...`.

### CODE REVIEW

File: extensions/amp-a4a/0.1/test/test-refresh.js [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        expect(refreshManager.visibilityTimeoutId_).to.be.ok;
+        expect(refreshManager.state_).to.equal('view_pending');
+        ioEntry.intersectionRatio = 0;
+        refreshManager.ioCallback_([ioEntry]);
+        expect(refreshManager.visibilityTimeoutId_).to.be.null;
+        expect(refreshManager.state_).to.equal('initial');
+      });
+    });
+
+    it('should execute the refresh event correctly', () => {
```
Comment (jonkeller) [inline extensions/amp-a4a/0.1/test/test-refresh.js]: Looks like nothing above this line is changed in this file - please correct me if I am wrong.

File: extensions/amp-a4a/0.1/refresh-intersection-observer-wrapper.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    /** @private @const {!../../../src/service/viewport/viewport-impl.Viewport} */
+    this.viewport_ = baseElement.getViewport();
+  }
+
+  /**
+   * Begin observing the given element.
+   * @param {!Element} element
+   */
+  observe(element) {
+    const refreshId = element.getAttribute('data-amp-ad-refresh-id');
```
[the remark this answers is not in the document]
Comment (glevitzky) [reply extensions/amp-a4a/0.1/refresh-intersection-observer-wrapper.js]: It is exported in refresh-manager.js as a constant, but unfortunately, I can't import it without creating a cyclical dependency. I'll add a comment.

File: extensions/amp-a4a/0.1/test/test-refresh.js (hunk 1 above)
Comment (glevitzky) [reply extensions/amp-a4a/0.1/test/test-refresh.js]: That's correct.

File: extensions/amp-a4a/0.1/refresh-manager.js [hunk 3]
```diff
@@ -310,8 +312,7 @@ export class RefreshManager {
     return new Promise(resolve => {
       this.refreshTimeoutId_ = this.timer_.delay(() => {
         this.state_ = RefreshLifecycleState.INITIAL;
-        this.getIntersectionObserverWithThreshold_(
-            this.config_.visiblePercentageMin).unobserve(this.element_);
+        this.unobserve();
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/refresh-manager.js]: Why do we need to unobserve and then observe again afterall the element never changes as its the amp-ad element, right?  Wouldn't it be simpler to just leave the observe in place and then start the timeout when you first enter viewport after a refresh?  We can discuss outside of the PR

File: extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js [hunk 4]
```diff
@@ -861,6 +861,9 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
           return false;
         }
     }
+    if (this.refreshManager_) {
```
[the remark this answers is not in the document]
Comment (glevitzky) [reply extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js]: That's fine. If I recall correctly, we decided not to keep any reference to RefreshManager in AmpA4A to minimize bloat. So this is more-or-less a consciously made trade-off.

File: extensions/amp-a4a/0.1/refresh-intersection-observer-wrapper.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+  /**
+   * Cease observing the given element.
+   * @param {!Element} element
+   */
+  unobserve(element) {
+    // We need to call 'tick' to update current host viewport state, otherwise
+    // the next time we call 'observe', the viewport state might be stale, and
+    // indicate that the element is in the viewport when it's not.
+    this.intersectionObserver_.tick(this.viewport_.getRect());
```
[the remark this answers is not in the document]
Comment (glevitzky) [reply extensions/amp-a4a/0.1/refresh-intersection-observer-wrapper.js]: If you call `tick` before `observe`, it essentially no-ops.

File: extensions/amp-a4a/0.1/test/test-refresh.js (hunk 1 above)
Comment (rsimha) [reply extensions/amp-a4a/0.1/test/test-refresh.js]: @glevitzky This test is failing on master. Can you investigate and skip / fix the test?

https://travis-ci.org/ampproject/amphtml/jobs/373920148

Comment (glevitzky) [reply extensions/amp-a4a/0.1/test/test-refresh.js]: Sorry about that. Fix is here: https://github.com/ampproject/amphtml/pull/15019/.

**Pull request merged**